### DISCUSSION TITLE
Expose Payment REST API

### DISCUSSION THREAD

**Discussion by cpoisnel** [pretix/pretix#2663] [Q&A] [answered]
Hi,

Linked to #654 "Writable API", I would like to know if there are plans to expose endpoints to configure the payment (those provided as standard by Pretix for example): 

* Ability to configure manual payment
* Ability to configure gift cards
* ...

We are struggling with how Django Rest Framework works and the ViewSet (linked to providers and models) and would like to know if there is already an example of API design applicable on this type of plugin.

Thanks in advance for your feedback,

**Comment by raphaelm** [comment, marked as answer]
Generally, most (if not all) payment providers store their configuration in the "settings" data store.

Since 3.6, the settings store is **partly** accessible through the API.
https://docs.pretix.eu/en/latest/api/resources/events.html#event-settings

This mechanism, however, is **opt-in** for every setting for safety reasons.
https://docs.pretix.eu/en/latest/development/api/general.html#pretix.base.signals.api_event_settings_fields

We did not add any payment provider's settings to the API so far for three reasons:

* Many payment providers require extended validation, for example it should not be possible to enable PayPal payments if no PayPal API key is set. It's a lot of work and risk to maintain this logic for the API, and it's hard for API users to understand and use.

* Many payment providers require configuration steps that go beyond what a simple read/write API can do, e.g. if there needs to be some kind of OAuth handshake.

* Payment settings usually contain *secrets*, such as API keys for a payment gateway. We consider it a security feature that you can never get these secrets out of pretix again, so we'd first need a mechanism for "write-only settings" that are not readable again.

I belive for the two examples you picked, manual payment and gift cards, neither of these reasons apply and it wouldn't be too hard to add these to the API.

But for almost every other payment provider… oof, that's gonna be tough to build and maintain in a nice way.

**Comment by cpoisnel** [comment]
Thank you for your feedback.

> I belive for the two examples you picked, manual payment and gift cards, neither of these reasons apply and it wouldn't be too hard to add these to the API.

I hope so :+1:
